core/src/main/scala/com/acervera/osm4scala/model/RelationEntity.scala

```diff
@@ -25,35 +25,93 @@
+                          changeset: Option[Long],
+                          uid: Option[Int],
+                          user: Option[String],
+                          visible: Option[Boolean]) extends OSMEntity {
 
   override val osmModel: OSMTypes.Value = OSMTypes.Relation
 
+  def apply(id: Long,
+            relations: Seq[RelationMemberEntity],
+            tags: Map[String, String]): RelationEntity = {
+    RelationEntity(id, relations, tags,
+      None, None, None, None, None, None)
+  }
+
+  override def toString: String = {
+    s"Relation id: ${id}, " +
+      s"relations: ${relations}, " +
+      s"tags: ${tags.toList}, " +
+      s"version: ${version.getOrElse("None")}," +
+      s"timestamp: ${timestamp.getOrElse("None")}, " +
+      s"changeset: ${changeset.getOrElse("None")}, " +
+      s"uid: ${uid.getOrElse("None")}, " +
+      s"user: ${user.getOrElse("None")}, " +
+      s"visible: ${visible.getOrElse("None")}\n"
+  }
+
 }
 
 object RelationEntity {
 
-  def apply(osmosisStringTable: StringTable, osmosisRelation: Relation): RelationEntity = {
+  val DEFAULT_DATE_GRANULARITY: Int = 1000
+
+  def apply(osmosisRelation: Relation, osmosisStringTable: StringTable): RelationEntity = {
+    apply(osmosisRelation, osmosisStringTable, Option[Int](DEFAULT_DATE_GRANULARITY))
+  }
+
+  def apply(osmosisRelation: Relation,
+            osmosisStringTable: StringTable,
+            dateGranularity: Option[Int]): RelationEntity = {
+
+    val _dateGranularity: Int = dateGranularity.getOrElse(DEFAULT_DATE_GRANULARITY)
 
     // Calculate tags using the StringTable.
-    val tags = (osmosisRelation.keys, osmosisRelation.vals).zipped.map { (k, v) =>
-      osmosisStringTable.s(k).toString("UTF-8") -> osmosisStringTable.s(v).toString("UTF-8")
+    val tags: Map[String, String] = (osmosisRelation.keys, osmosisRelation.vals).zipped.map { (k, v) =>
+      osmosisStringTable.s(k).toString(DecompressUtils.STRING_ENCODER) -> osmosisStringTable.s(v).toString(DecompressUtils.STRING_ENCODER)
     }.toMap
 
     // Decode members references in stored in delta compression.
-    val members = osmosisRelation.memids.scanLeft(0L) { _ + _ }.drop(1)
+    val members: Seq[Long] = osmosisRelation.memids.scanLeft(0L) { _ + _ }.drop(1)
+
+    val optionalInfo: Option[Info] = osmosisRelation.info
 
     // Calculate relations
-    val relations = (members, osmosisRelation.types, osmosisRelation.rolesSid).zipped.map { (m, t, r) =>
-      RelationMemberEntity(m, RelationMemberEntityTypes(t.value), osmosisStringTable.s(r).toString("UTF-8"))
+    val relations: Seq[RelationMemberEntity] = (members, osmosisRelation.types, osmosisRelation.rolesSid).zipped.map { (m, t, r) =>
+      RelationMemberEntity(m, RelationMemberEntityTypes(t.value), osmosisStringTable.s(r).toString(DecompressUtils.STRING_ENCODER))
     }
 
-    new RelationEntity(osmosisRelation.id, relations, tags)
+    val version: Option[Int] = optionalInfo.filter(_.version.isDefined).map(_.version.get)
```

Comment:
This code duplication (check line wayEntity:95) is smell code and could be related to a design problem.
The same for the following lines of code.

Reply:
Yeah, I will extract and rearrange it